File: src/waypalm-site/src/main/java/waypalm/site/web/mvc/oauth/OAuth2Flow.java
package waypalm.site.web.mvc.oauth;

import org.springframework.social.connect.Connection;
import org.springframework.social.connect.support.OAuth2ConnectionFactory;
import org.springframework.social.oauth2.AccessGrant;
import org.springframework.social.oauth2.GrantType;
import org.springframework.social.oauth2.OAuth2Operations;
import org.springframework.social.oauth2.OAuth2Parameters;

import javax.security.sasl.AuthenticationException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class OAuth2Flow extends OAuthFlow<OAuth2ConnectionFactory<?>> {
    public OAuth2Flow(boolean initiationRequired, OAuth2ConnectionFactory<?> factory) {
        super(initiationRequired, factory);
    }

    @Override
    public void initiateAuthentication(HttpServletRequest request, HttpServletResponse response) throws IOException {
        OAuth2Operations operations = getConnectionFactory().getOAuthOperations();
        OAuth2Parameters parameters = new OAuth2Parameters();
        parameters.setRedirectUri(getCallbackUrl());
        response.sendRedirect(operations.buildAuthenticateUrl(GrantType.AUTHORIZATION_CODE, parameters));
    }

    @Override
    public Connection<?> attemptAuthentication(HttpServletRequest request) throws AuthenticationException {
        String code = request.getParameter("code");
        AccessGrant accessGrant = getConnectionFactory().getOAuthOperations().exchangeForAccess(code, getCallbackUrl(), null);
        return getConnectionFactory().createConnection(accessGrant);
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/web/mvc/oauth/OAuth2Flow.kt
package waypalm.site.web.mvc.oauth

import org.springframework.social.connect.support.OAuth2ConnectionFactory
import javax.servlet.http.HttpServletRequest
import org.springframework.social.connect.Connection
import javax.servlet.http.HttpServletResponse
import org.springframework.social.oauth2.OAuth2Parameters
import org.springframework.social.oauth2.GrantType

public class OAuth2Flow(
        initiationRequired: Boolean,
        factory: OAuth2ConnectionFactory<*>,
        callbackUrl: String
)
: OAuthFlow<OAuth2ConnectionFactory<*>>(initiationRequired, factory, callbackUrl)
{
    public override fun attemptAuthentication(request: HttpServletRequest): Connection<*>? {
        val code = request.getParameter("code");
        val accessGrant = connectionFactory.getOAuthOperations()!!.exchangeForAccess(code, callbackUrl, null);
        return connectionFactory.createConnection(accessGrant);
    }

    public override fun initiateAuthentication(request: HttpServletRequest, response: HttpServletResponse) {
        var operations = connectionFactory.getOAuthOperations();
        val parameters = OAuth2Parameters();
        parameters.setRedirectUri(callbackUrl);
        response.sendRedirect(operations!!.buildAuthenticateUrl(GrantType.AUTHORIZATION_CODE, parameters));
    }
}